[MRG+1] Fix cross_val_predict behavior for binary classification in decision_function (Fixes #9589)

reiinakano opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->
Fixes #9589 

#### What does this implement/fix? Explain your changes.
I figured that the purpose of this code (original)

```python
if method in ['decision_function', 'predict_proba', 'predict_log_proba']:
        n_classes = len(set(y))
        predictions_ = np.zeros((X_test.shape[0], n_classes))
        if method == 'decision_function' and len(estimator.classes_) == 2:
                predictions_[:, estimator.classes_[-1]] = predictions
        else:
            predictions_[:, estimator.classes_] = predictions
        predictions = predictions_
```

is to handle cases where the cross-validation done in `cross_val_predict` doesn't properly stratify the classes i.e. one fold in the CV may not have all classes present.

However, it misses an edge case for `decision_function` when `len(estimator.classes_) == 2` (train split has only 2 classes) and `n_classes == 2` (total data has only 2 classes). It wrongly assumes that the output will be of shape `(n_samples, n_classes)`, when in fact, this is only true for `decision_function` if `n_classes > 2`. For the correct behavior, we must first check if the total number of classes is greater than 2, and if it is not, we must stick to an output shape of `(n_samples,)`.

#### Any other comments?

While this PR fixes the case raised in #9589, it still doesn't take into account the fact that for `sklearn.svm.SVC`, `decision_function` does not always follow the shape `(n_samples, n_classes)` or `(n_samples,)`. If `decision_function_shape` is not set to `ovr`, the shape is `(n_samples, n_classes * (n_classes-1) / 2)`, and if different train splits have different numbers of classes (not stratified), this inconsistency can't be fixed by the approach currently taken by `cross_val_predict`.

I would also like to point out my personal opinion that `cross_val_predict` should probably just throw an error if the classes are not stratified properly (by this I mean some classes are entirely absent in some splits) instead of handling it automatically as it tries to do now. Metrics like log loss and ROC might be affected if for some splits, entire columns are zeroed out. Idk, it just feels inconsistent. Of course, you guys have the final say on this.

What I mean by inconsistent:

```python
from sklearn.model_selection import cross_val_predict, KFold
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression

X, y = load_iris(True)
cross_val_predict(LogisticRegression(), X, y, cv=KFold().split(X), method='predict_proba')
```

returns (array cut down in size)

```
array([[  0.00000000e+00,   9.99987841e-01,   1.21585683e-05],
       [  0.00000000e+00,   9.99963190e-01,   3.68104260e-05],
       [  0.00000000e+00,   9.99970224e-01,   2.97757119e-05],
       [  0.00000000e+00,   9.99978481e-01,   2.15185386e-05],
       [  0.00000000e+00,   9.99993136e-01,   6.86428459e-06],
       [  1.20030529e-01,   0.00000000e+00,   8.79969471e-01],
       [  1.13183668e-01,   0.00000000e+00,   8.86816332e-01],
       [  6.75506225e-02,   0.00000000e+00,   9.32449377e-01],
       [  1.07273496e-01,   0.00000000e+00,   8.92726504e-01],
       [  1.26854747e-01,   0.00000000e+00,   8.73145253e-01],
       [  1.16881976e-01,   0.00000000e+00,   8.83118024e-01],
       [  2.45077442e-04,   9.99754923e-01,   0.00000000e+00],
       [  2.70475201e-04,   9.99729525e-01,   0.00000000e+00],
       [  6.59523734e-04,   9.99340476e-01,   0.00000000e+00],
       [  4.07574999e-04,   9.99592425e-01,   0.00000000e+00],
       [  2.42451670e-03,   9.97575483e-01,   0.00000000e+00],
       [  2.03959503e-03,   9.97960405e-01,   0.00000000e+00]])
```
Notice how there are "zeroed out" columns.

And even worse:

```python
from sklearn.model_selection import cross_val_predict, KFold
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression

X, y = load_iris(True)
cross_val_predict(LogisticRegression(), X, y, cv=KFold().split(X), method='decision_function')
```
returns (array cut down in size)
```
array([[  0.        ,   0.        , -11.31746426],
       [  0.        ,   0.        , -10.20969263],
       [  0.        ,   0.        , -10.42178776],
       [  0.        ,   0.        ,  -9.60397656],
       [  0.        ,   0.        , -11.30005126],
       [  0.        ,   0.        , -11.1906238 ],
       [  0.        ,   0.        , -10.05510072],
       [  0.        ,   0.        , -10.74657422],
       [  0.        ,   0.        ,  -9.20295991],
       [  0.        ,   8.3136912 ,   0.        ],
       [  0.        ,   6.77281268,   0.        ],
       [  0.        ,   7.79874366,   0.        ],
       [  0.        ,   7.29615134,   0.        ],
       [  0.        ,   7.9199709 ,   0.        ],
       [  0.        ,   9.163118  ,   0.        ],
       [  0.        ,   5.8858718 ,   0.        ],
       [  0.        ,   6.37425957,   0.        ],
       [  0.        ,   6.66261924,   0.        ],
       [  0.        ,   6.19296233,   0.        ]])
```
which is hardly of use to anyone.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

reiinakano committed:
fix cross_val_predict for binary classification in decision_function

jnothman commented:
I think better not to error in the context of using this in ClassifierChain or a stacking meta-estimator.

reiinakano commented:
> I think better not to error in the context of using this in ClassifierChain or a stacking meta-estimator.

Hmm, you're right. One note though, I can't speak for `ClassifierChain` but for stacking, these zeroed out columns would mess with how the secondary learner learns and I wouldn't trust the results. Perhaps display a warning?

> Yes, perhaps 0 isn't the best pick, particularly for predict_proba. NaN?

NaN makes it clearer that these columns weren't considered at all for this particular split, but wouldn't that break things? For instance, I frequently use `np.argmax(cross_val_predict(..., method='predict_proba))` to recover the true labels for calculating accuracy, etc. Of course, I always make sure my folds are stratified so I'm probably not going to face it, but still.. Also, wouldn't that break ClassifierChains and stacking meta-estimators?

jnothman commented:
yes, nan would probably break it all. a warning is probably a good idea.
But first we need to fix the regression.

On 21 Aug 2017 11:55 pm, "Reiichiro Nakano" <notifications@github.com>
wrote:

> I think better not to error in the context of using this in
> ClassifierChain or a stacking meta-estimator.
>
> Hmm, you're right. One note though, I can't speak for ClassifierChain but
> for stacking, these zeroed out columns would mess with how the secondary
> learner learns and I wouldn't trust the results. Perhaps display a warning?
>
> Yes, perhaps 0 isn't the best pick, particularly for predict_proba. NaN?
>
> NaN makes it clearer that these columns weren't considered at all for this
> particular split, but wouldn't that break things? For instance, I
> frequently use np.argmax(cross_val_predict(..., method='predict_proba))
> to recover the true labels for calculating accuracy, etc. Of course, I
> always make sure my folds are stratified so I'm probably not going to face
> it, but still.. Also, wouldn't that break ClassifierChains and stacking
> meta-estimators?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9593#issuecomment-323749514>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz63bP5qekOvF21sYAcB9DUrp3Nc5mks5saYw_gaJpZM4O9TZp>
> .
>

jnothman reviewed (commented):
Could you please add a non-regression test.

reiinakano committed:
Add unit tests

reiinakano committed:
Add unit tests

reiinakano committed:
Add unit tests

reiinakano commented:
I've added the necessary unit tests.

Although at this point, I'd like to point out there's another regression for the case of SVC with `decision_function_shape=ovo`

Code to test:
```python
from sklearn.datasets import load_digits
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_predict

X, y = load_digits(return_X_y=True)
cross_val_predict(SVC(kernel='linear', decision_function_shape='ovo'), X, y, method='decision_function').shape
```

**Expected Results**
In previous versions, this is `(1797, 45)`

**Actual Results**
```
ValueError: shape mismatch: value array of shape (602,45) could not be broadcast to indexing result of shape (10,602)
```

This error is due to the shape of an SVC decision function with `decision_function_shape` set to `'ovr'` which is `(n_samples, n_classes * (n_classes-1) / 2)`

Should I include the fix in this PR?

jnothman commented:
Thanks. Maybe fix that in a separate PR unless they're going to step on
each others' toes.


On 22 Aug 2017 3:14 pm, "Reiichiro Nakano" <notifications@github.com> wrote:

I've added the necessary unit tests.

Although at this point, I'd like to point out there's another regression
for the case of SVC with decision_function_shape=ovo

Code to test:

from sklearn.datasets import load_digitsfrom sklearn.svm import
SVCfrom sklearn.model_selection import cross_val_predict

X, y = load_digits(return_X_y=True)
cross_val_predict(SVC(kernel='linear', decision_function_shape='ovo'),
X, y, method='decision_function').shape

*Expected Results*
In previous versions, this is (1797, 45)

*Actual Results*

ValueError: shape mismatch: value array of shape (602,45) could not be
broadcast to indexing result of shape (10,602)

This error is due to the shape of an SVC decision function with
decision_function_shape set to 'ovr' which is (n_samples, n_classes *
(n_classes-1) / 2)

Should I include the fix in this PR?

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/scikit-learn/scikit-learn/pull/9593#issuecomment-323921155>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAEz6wWf_xiKuaUe2MepWSuyunoj1sjYks5samOegaJpZM4O9TZp>
.

jnothman reviewed (commented):
LGTM

reiinakano committed:
better fix

reiinakano committed:
fix conflict

reiinakano committed:
Merge branch 'master' into fix-cross-val-predict-decision-function

reiinakano commented:
Hi @jnothman, I changed this back to WIP since there was a conflict with a recent merge. I also figured out a better fix that also fixes the SVC "ovo" regression.

jnothman commented:
Okay. Let me know when it's MRG, i.e. ready for review.

reiinakano committed:
fix broken

reiinakano committed:
Merge remote-tracking branch 'origin/fix-cross-val-predict-decision-function' into fix-cross-val-predict-decision-function

reiinakano committed:
only calculate n_classes if one of 'decision_function', 'predict_proba', 'predict_log_proba'

reiinakano commented:
@jnothman Ready for review

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -732,12 +732,13 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
-        else:
-            predictions_[:, estimator.classes_] = predictions
-        predictions = predictions_
+        if not n_classes == len(estimator.classes_):
```
not ... == -> ... !=

jnothman reviewed (commented):
Did you mean to add a test for the OvO case?

reiinakano committed:
add test for SVC ovo in cross_val_predict

reiinakano committed:
flake8 fix

reiinakano commented:
Yup, sorry. Added unit test and addressed comments.

jnothman reviewed (commented):
LGTM, thanks!

amueller commented:
A solution to some more of this will be in #9532 but that will probably still take a while.

amueller commented on sklearn/model_selection/_validation.py:
```diff
@@ -732,12 +732,13 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
-        else:
-            predictions_[:, estimator.classes_] = predictions
-        predictions = predictions_
+        if n_classes != len(estimator.classes_):
```
Should we raise a warning in this case?

amueller replied:
This implicitly checks whether ``n_classes > 2``, right? If there was an estimator that would work for 1 class, the bug would still appear, right? Should we add an explicit check that ``n_classes > 2`` to guard against that?

amueller replied:
I don't see where the ``ovo`` case is addressed. We should also raise an error if ``predictions.shape[1] != len(estimator.n_classes_)``, right?

The test for this case is in ``test_cross_val_predict_with_method``. I suggest we add a test with ``SVC(decision_function_shape='ovo')`` and make sure a helpful error is thrown.

reiinakano replied:
@amueller I agree that a warning should be raised in this case.

I don't follow your statement about `n_classes > 2`. Which bug are you referring to? Could you give a specific example?

For your third point, what I've fixed here is the following regression of `cross_val_predict` behavior when `decision_function_shape=ovo`

Code to test:
```python
from sklearn.datasets import load_digits
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_predict

X, y = load_digits(return_X_y=True)
cross_val_predict(SVC(kernel='linear', decision_function_shape='ovo'), X, y, method='decision_function').shape
```

**Expected Results**
In previous versions, this is `(1797, 45)`

**Actual Results**
```
ValueError: shape mismatch: value array of shape (602,45) could not be broadcast to indexing result of shape (10,602)
```

I've also included the test for the `ovo` case in Line 792-798 in `test_validation`. Should I move this somewhere else?

I also acknowledge that the `ovo` case still throws an error if the folds are not properly stratified. However, this has always been the case even in previous versions, and is due to the special shape of `decision_function` for `ovo`.

amueller replied:
> I also acknowledge that the ovo case still throws an error if the folds are not properly stratified. However, this has always been the case even in previous versions, and is due to the weird shape of decision_function for ovo.

That was the case I meant. I just wanted a better error in this case, and I want a test for that error.

amueller reviewed (commented):
I think the fix is good but I'd like to see a more explicit test for ``SVC(decision_function_shape='ovo')`` and some more guards that what we do is actually correct.

amueller commented:
> I don't follow your statement about n_classes > 2. Which bug are you referring to? Could you give a specific example?

```python
from sklearn.linear_model import RidgeClassifier
from sklearn.datasets import load_iris
from sklearn.model_selection import cross_val_predict, KFold

X, y = load_iris(return_X_y=True)
X = X[:100]
y = y[:100]
res = cross_val_predict(RidgeClassifier(), X, y, method='decision_function', cv=KFold(2))
```
> ValueError: shape mismatch: value array of shape (50,) could not be broadcast to indexing result of shape (1,50)

amueller commented:
This is the case when some classifiers have 1 class and some have 2 classes.

reiinakano commented:
Got it. Didn't know some classifiers could be fit on a single class. So that's *another* regression. How do you want to handle this case? It sounds similar to the case `if method == 'decision_function' and len(estimator.classes_) == 2`, which is currently handled by doing `predictions_[:, estimator.classes_[-1]] = predictions`.

If that's fine, I think the fix is to simply change the former line to `if method == 'decision_function' and len(estimator.classes_) <= 2`.

This would fix the errors, but frankly, I'd still prefer a stern warning that something went wrong during cross-validation using `cross_val_predict`.

For the ovo case, should I include that in this PR as well? I didn't include it previously since this PR primarily fixes regressions and I didn't want to include a new "feature".

amueller commented:
Not sure if the 1 class thing is a regression or was always broken ;)
And I think I'd like the error / test for OVO in this PR.

Sorry, I don't understand your proposal for the 1class / 2class case. The result in the 2 class case should be 1d, not 2d. That's the original bug this is trying to solve, right?

reiinakano commented:
It's a regression. The code you wrote above only breaks on 0.19. :)

> Sorry, I don't understand your proposal for the 1class / 2class case. The result in the 2 class case should be 1d, not 2d. That's the original bug this is trying to solve, right?

The case I'm talking about is when the original dataset has >2 classes, and a particular fold is <=2 *and* the method is `decision_function`. The original bug is still fixed by the `if n_classes != len(estimator.classes_):` check.

amueller commented:
Ah. My breaking example is with the dataset having two classes, and some folds have one class.

reiinakano commented:
Okay, I finally see your original point here

>This implicitly checks whether n_classes > 2, right? If there was an estimator that would work for 1 class, the bug would still appear, right? Should we add an explicit check that n_classes > 2 to guard against that?

You're right. It's not a regression, but another edge case that needs to be fixed. Yes, we should add an explicit check that `n_classes > 2` but only if the method used is `decision_function`.

reiinakano committed:
fix case of ovo and imbalanced folds for binary classification

reiinakano committed:
Merge branch 'master' into fix-cross-val-predict-decision-function

reiinakano committed:
change assert_raises to assert_raise_message for ovo case

reiinakano committed:
fix flake8 linetoo long

reiinakano commented:
@amueller I've added both the error for OvO and the fix for a dataset having two classes, and some folds having one class. Added unit tests for both.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -727,12 +727,23 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+            else:
+                predictions_[:, estimator.classes_] = predictions
+            if method == 'decision_function' and n_classes == 2:
+                predictions_ = predictions_[:, 1]
```
this line does not appear to be covered by tests.

amueller replied:
I'm a little bit confused by this if else sequence. Why do we need the first if in addition to the second if? doesn't that cover all cases?

reiinakano replied:
Let's continue the conversation here. Sorry about that.

> Is there a case with n_classes == 2 where len(estimator.classes_) > 2? It looks to me like these two ifs are always either both true or neither but I might be missing something.

No but there are cases when `n_classes != 2` and `len(estimator.classes_) <= 2`

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -776,6 +777,44 @@ def split(self, X, y=None, groups=None):
 
     assert_raises(ValueError, cross_val_predict, est, X, y, cv=BadCV())
 
+    X, y = make_classification(n_classes=2, n_samples=50)
```
Can we make this a separate function, like `test_cross_val_predict_decision_function_shape`?

Although I'd also like `predict_proba` similarly tested.

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -776,6 +777,44 @@ def split(self, X, y=None, groups=None):
+                              method='decision_function')
+    assert_equal(preds.shape, (150, 3))
+
+    X = X[:100]
```
Please add a comment on how this test differs from the 2-class one above.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -727,12 +727,23 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+        if n_classes != len(estimator.classes_):
+            if len(predictions.shape) == 2 and \
+                    predictions.shape[1] != len(estimator.classes_):
+                raise ValueError('Shape of output predictions does not '
```
I think we need a comment here to state what kind of a case this deals with.

jnothman reviewed (commented):
Otherwise, still LGTM

reiinakano committed:
add comments and clearer tests

reiinakano commented:
@jnothman Addressed comments. Failure in Appveyor appears from an unrelated test.

reiinakano committed:
Merge remote-tracking branch 'upstream/master' into fix-cross-val-predict-decision-function

reiinakano commented:
I'm not sure why an unrelated test in `sklearn.neighbors` is failing in Appveyor when the only difference in the commit that caused the failure is a few comments and additional unit tests.

jnothman commented:
That appears to be a very rare failure at random. I'll fix that test and restart appveyor.

reiinakano committed:
Merge remote-tracking branch 'upstream/master' into fix-cross-val-predict-decision-function

reiinakano commented:
Finally green. Thanks! Ready for review.

amueller commented on sklearn/model_selection/_validation.py:
```diff
@@ -727,12 +727,27 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
-            predictions_[:, estimator.classes_] = predictions
-        predictions = predictions_
+        if n_classes != len(estimator.classes_):
+            if len(predictions.shape) == 2 and \
```
``predictions.ndim``? ``predictions.ndim == 2`` not the case for multioutput multi-class, I guess?

amueller commented on sklearn/model_selection/_validation.py:
```diff
@@ -727,12 +727,27 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+                # does not match the number of classes used to train
+                # it with. This case is found when sklearn.svm.SVC is
+                # set to `decision_function_shape='ovo'`.
+                raise ValueError('Shape of output predictions does not '
```
I think it should be ``Output shape of {}...``.format(method)``. It's still not super clear but I don't have a great suggestion either.

amueller commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -777,6 +778,64 @@ def split(self, X, y=None, groups=None):
+    y = y[:100]
+    preds = cross_val_predict(RidgeClassifier(), X, y,
+                              method='decision_function', cv=KFold(2))
+    assert_equal(preds.shape, (100,))
```
Can we somehow check that the signs here are correct, not just the shape?

reiinakano replied:
I can't seem to comment on the review just above this (Github bug?), so I'll put my reply in here.

> I'm a little bit confused by this if else sequence. Why do we need the first if in addition to the second if? doesn't that cover all cases?

There is a lot of special casing here so not sure what you mean by "first if" and "second if" but I'll try to explain anyway.

```python
if method == 'decision_function' and n_classes == 2:
    predictions_ = predictions_[:, 1]
```

pertains to the case you mentioned "when some classifiers have 1 class and some have 2 classes." Without this line, the output shape would be `(n_samples, 2)` instead of `(n_samples,)`. This is only applicable to `decision_function`. It forces `predictions_` into the correct shape.

If the first if you pertain to is 
```python
if method == 'decision_function' and len(estimator.classes_) <= 2:
    predictions_[:, estimator.classes_[-1]] = predictions
```
then this is for handling when a particular CV fold has <2 classes, and we need to handle it properly when building `predictions_`. 

> Can we somehow check that the signs here are correct, not just the shape?

Not sure what you mean by signs.

amueller replied:
Is there a case with ``n_classes == 2`` where ``len(estimator.classes_) > 2``? It looks to me like these two ``if``s are always either both true or neither but I might be missing something.

By sign, I mean that in the two-class case, the meaning of the decision function changes depending on whether the one class is the positive class or the negative class, right? If the decision function in the one-class case is always 0, I guess it doesn't make a difference, but I haven't checked that.

reiinakano replied:
> Is there a case with n_classes == 2 where len(estimator.classes_) > 2? It looks to me like these two ifs are always either both true or neither but I might be missing something.

No but there are cases when `n_classes != 2` and `len(estimator.classes_) <= 2`

> By sign, I mean that in the two-class case, the meaning of the decision function changes depending on whether the one class is the positive class or the negative class, right? If the decision function in the one-class case is always 0, I guess it doesn't make a difference, but I haven't checked that.

For `RidgeClassifier`, it's actually -1. I'm not sure what that means and it's not a regression so I didn't attempt to fix it here. That said, I'm not really sure what a "correct" implementation would look like.

amueller replied:
I'll double check. Can you comment on the first part? Is there a case when only one of the two if's is true?

amueller replied:
Can you provide the code you used for RidgeClassifier?

reiinakano replied:
If you check the "Files changed" tab, you'll actually see my comment on the first part. Weird Github bug doesn't show it here I guess.. So i'll copy it here.

> Is there a case with n_classes == 2 where len(estimator.classes_) > 2? It looks to me like these two ifs are always either both true or neither but I might be missing something.

No but there are cases when `n_classes != 2` and `len(estimator.classes_) <= 2`

reiinakano replied:
Here's the code I used for `RidgeClassifier`. 

```python
from sklearn.datasets import load_iris
X,y = load_iris(True)
from sklearn.linear_model import RidgeClassifier
from sklearn.model_selection import cross_val_predict
cross_val_predict(RidgeClassifier(), X[:20], y[:20], method='decision_function')
```
```
array([-1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.,
       -1., -1., -1., -1., -1., -1., -1.])
```
this is v0.18

amueller reviewed (commented):
Sorry for dragging this on, but I find the code not super obvious, and it's not entirely clear to me that the signs are right in the 2 class decision function case (and I don't see a test for that).

amueller commented:
Ok, I think if one fold only contains one class, we should probably just error becaus it's a bit unclear to me what is supposed to happen. And this is such a weird edge-case that I don't think we should hold 0.19.1 for that.

For the if-else thing in the decision function case: maybe just add some comments? right now I find the code hard to read. But seems ok to me.

reiinakano committed:
improve comments and error message for OvO

reiinakano commented:
@amueller I've addressed your comments. So, ready for review again.

The test failures are due to something unrelated again..

reiinakano committed:
fix .format error with L

reiinakano commented:
@amueller addressed your comments. Ready for review again.

The test failures seem to be due to something unrelated again...

jnothman commented:
Add a blank commit and the failures should go away due to a patch in master

reiinakano committed:
Merge branch 'master' into fix-cross-val-predict-decision-function

reiinakano commented:
So.. something failed. But again, I have no idea what...

lesteve commented:
You can see that this is lgtm failing (google it if you want to know more what lgtm does). For some reason it has done that recently, not sure why but you don't need to worry about this.

lesteve reviewed (commented):
You probably need an entry in whats_new.

I haven't had time to look into the main change but I have a few comments.

reiinakano commented on sklearn/model_selection/_validation.py:
```diff
@@ -727,12 +727,35 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+                # does not match the number of classes used to train
+                # it with. This case is found when sklearn.svm.SVC is
+                # set to `decision_function_shape='ovo'`.
+                raise ValueError('Output shape (-1, {}) of {} does not '
```
It's supposed to mean that only the second value in the shape matters. I initially just put in the shape but the test was failing because `.format()` would put in either `(10, 10)` or `(10L, 10L)`.

 Do you think something like "Inner dimension of output shape {}" would be better?

lesteve replied:
That's what I half-guessed but this is very confusing I am afraid. 

> I initially just put in the shape but the test was failing because .format() would put in either (10, 10) or (10L, 10L).

So you made a worse error message just to get the test to pass. I think you need to set your priorities straight ;-) ! You can use assert_raises_regex with a regex. Use git `grep L?` for an example. The important thing is to have a good error message.

reiinakano replied:
Ah, my bad. I didn't know about assert_raises_regex. Will get on a fix immediately!

reiinakano commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -777,6 +778,64 @@ def split(self, X, y=None, groups=None):
+                              method='decision_function')
+    assert_equal(preds.shape, (50,))
+
+    X, y = load_iris(return_X_y=True)
```
No reason, except that it flows naturally into the next test where I need a balanced binary classification dataset sorted by label. Should I use another `make_classification` call here?

lesteve replied:
Yeah I saw that after I commented, I think it's fine, leave it as it is.

reiinakano committed:
use assert_raises_regex for better error message

jnothman commented:
I'll add a what's new entry when throwing it into 0.19.1, to avoid having it incorrectly in 0.20

reiinakano committed:
Merge branch 'master' into fix-cross-val-predict-decision-function

lesteve commented:
I looked at this for a bit, my main comment: with your change it seems like we are not very consistent  when we ave a training fold which does not have all the training labels present in the complete dataset:
* if prediction.ndim == 2, which I think for method='decision_function' means `len(est.classes_) >= 3`, we raise an error saying use a proper stratification technique
* if prediction.ndim < 2, which I think for method='decision_function' means `len(est.classes_) <= 2`, we do not raise an error.

Also it seems like the fix for cross_val_predict in binary classification as reported in #9589 is actually moving the from the logic from https://github.com/scikit-learn/scikit-learn/pull/7889 in a `if n_classes != len(estimator.classes_)`. I would like to take a closer look at this to understand better whether this is the right thing to do.

reiinakano commented:
The error message is raised solely for a single special case, which is an `SVC` with `decision_function_shape='ovo'`. This is because the shape of `predictions` is neither `(n_samples, n_classes)` nor `(n_samples,)` (when `len(est.classes_) == 2`). To my knowledge, this does not happen anywhere else.

The fix for the original issue in #9589 is simply bypassing the special case handling when `n_classes == len(est.classes_)`. However, this has turned into a fix for a whole bunch of edge cases not covered in the original #7889.

> if predictions.ndim == 2, which I think for method='decision_function' means len(est.classes_) >= 3, we raise an error saying use a proper stratification technique

I don't think there is ever a case when `decision_function` returns `predictions.ndim == 2`.

lesteve commented:
> The error message is raised solely for a single special case, which is an SVC with decision_function_shape='ovo' with imbalanced splits.

You are right, I read the code too quickly, which to my defence is a bit tricky to follow.

I am not convinced we are doing the right thing when we don't have all the classes in the training fold (for example setting decision function to 0 in the missing classes does not seem right).
One way to avoid the problem would be to raise an error when we don't have all the classes in the training fold? Is there a reason why we don't do this?

reiinakano commented:
> One way to avoid the problem would be to raise an error when we don't have all the classes in the training fold?

I have thought about this too. A quote from my very first post above:

> I would also like to point out my personal opinion that cross_val_predict should probably just throw an error if the classes are not stratified properly (by this I mean some classes are entirely absent in some splits) instead of handling it automatically as it tries to do now. Metrics like log loss and ROC might be affected if for some splits, entire columns are zeroed out. Idk, it just feels inconsistent. Of course, you guys have the final say on this.

There is also some short discussion on this at the end of #8773.
__

> setting decision function to 0 in the missing classes does not seem right

Yup, I agree this does not make sense.

Perhaps what we could do is keep the special imbalanced handling code only for `predict_proba` and error out when it is imbalanced for `decision_function`.

jnothman commented:
The problem is that it does just happen in practice, when some classes are
infrequent. It might not make sense for decision_function to assume some
arbitrary value but it does make sense to zero out a column in
predict_proba.

reiinakano commented:
How about erroring out imbalanced classes in `decision_function` but zeroing out columns in `predict_proba` and `log_predict_proba`? Literally all the complication in this PR is because of the different weird ways `decision_function` behaves.

jnothman commented:
Essentially all we do in that case is make it harder for people to do the
same analysis on a highly imbalanced classification problem that they do
with a balanced one, which in turn makes it harder for people to provide
generic ML tools built on top of our library.

One example, again, is stacking.

jnothman commented:
I suppose that is, at least, more principled, and I can accept that. (And
by that I mean erroring for decision_function but not for
predict_{log_}proba).

Yes, these quirks and edge cases are annoying to deal with. But I feel like
sometimes it's our role as a library provider to handle them as best we
can, and raise an error only when the user has done something wrong or at
least very rare. I don't think this falls under very rare.

lesteve commented:
I read a bit the discussion in the related PRs, sorry it's taking me some time to catch up. Basically here is my summary:

* I would be a favour of a warning when the training fold does not contain all the classes. https://github.com/scikit-learn/scikit-learn/pull/8773#issuecomment-330362879 agrees with that as well.
* the fix for the wrong shape in cross_validation_predict with method=decision_function (which consists in only applying the logic from #7889 when `n_classes != n_classes_train`) is fine
* `method='decision_function'` has a number of edge cases. Raising an exception for `method='decision_function'` if `n_classes_train != n_classes` may be an easy way to avoid tackling all these special cases.
  1. SVC with ovo with uncommon number of columns for decision function `n_classes_train * (n_classes_train - 1) / 2` rather than `n_classes`. Tackled in this PR already.
  2. `n_classes_train <= 2` and `n_classes >= 3`, you get a 1d array as the output of decision_function and then it is not clear what to do with it. I would be in favour of raising an exception in this case.
  3. `n_classes_train == 1` and `n_classes == 2`. No idea what to do in this case. I don't know how common this is for classifier to be able to be fittable with a single class and have decision_function. I quickly looked using all_estimators and I found `['AdaBoostClassifier', 'LinearDiscriminantAnalysis',  'RidgeClassifier', 'RidgeClassifierCV']`.
* predictions should be initialized to 0 if `method='predict_proba'` but `-inf` (or maybe `np.finfo(X.dtype).min`) if `method='predict_log_proba'`.

reiinakano commented:
There is another case for `decision_function`

iv. `n_classes_train > 2` and `n_classes > 2` but `n_classes_train < n_classes`. Currently we set missing classes to 0.

lesteve commented:
> iv. n_classes_train > 2 and n_classes > 2 but n_classes_train != n_classes. Currently we set missing classes to 0. It has been mentioned earlier that this does not make sense. How should I handle this?

For me this is not really an edge case. This is the main case we are trying to solve. I was proposing to do this:

> predictions should be initialized to 0 if method='predict_proba' but -inf (or maybe np.finfo(X.dtype).min) if method='predict_log_proba'. 

For decision_function either we do the same as predict_log_proba or we decide that we always raise an exception, given that there are many edge cases for decision_function ...

reiinakano committed:
raise error in decision_function special cases. change predict_log_proba missing classes to minimum numpy value

reiinakano committed:
fix broken tests due to special cases of decision_function

reiinakano committed:
add modified test for decision_function behavior that does not trigger edge cases

reiinakano committed:
fix typos

reiinakano committed:
fix typos

reiinakano committed:
escape regex .

reiinakano committed:
escape regex .

reiinakano commented:
I think I've mostly sorted this out. For a non-special case of `decision_function` (`iv. n_classes_train > 2 and n_classes > 2 but n_classes_train != n_classes.`), I do the same as `predict_log_proba`. For the other special cases, there is an error message. 

Apologies for the confusing code previously, I should've done it this way in the first place, but I didn't expect `decision_function` to have so many edge cases.

lesteve commented on sklearn/model_selection/_validation.py:
```diff
@@ -727,12 +727,43 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+
+                predictions_ = np.full((_num_samples(X_test), n_classes),
+                                       np.finfo(predictions.dtype).min)
+                predictions_[:, estimator.classes_] = predictions
```
What changes for each method is the default value, you could simplify the code and use better variable names, e.g.  something like this in the non-error case:

```py
float_min = np.finfo(predictions.dtype).min
default_values = {'decision_function': float_min, 'predict_log_proba': float_min, 
    'predict_proba': 0}
predictions_for_all_classes = np.full((_num_samples(predictions), n_classes),
    default_values[method])
predictions_for_all_classes[:, estimator.classes_] = predictions
```

lesteve commented on sklearn/model_selection/_validation.py:
```diff
@@ -727,12 +727,43 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+                           ' in different folds. To fix this, use ' \
+                           'a cross-validation technique resulting' \
+                           ' in properly stratified folds'
+                if predictions.ndim == 2 and \
```
Parentheses rather than backslashes

lesteve commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1267,6 +1347,18 @@ def test_cross_val_predict_class_subset():
 
+    # Special test for decision_function. This makes sure not to trigger
+    # any of the numerous edge cases in decision_function
+    X = np.arange(200).reshape(100, 2)
```
I think you should keep decision_function in the loop above. Just find a dataset that works for decision_function. e.g. this one or one with many classes to avoid the edge cases that error.

lesteve reviewed (commented):
I think we still need a warning in any case when `n_classes_training != n_classes`.

I would still like feedback from some core devs about https://github.com/scikit-learn/scikit-learn/pull/9593#issuecomment-335178848, especially about the default value for missing classes for predict_log_proba and decision_function (personally I would favour `-inf` but maybe there are consumers of `cross_val_predict` which would raise an error with infinite values)

reiinakano committed:
address comments. one unaddressed comment

reiinakano committed:
simplify code

reiinakano committed:
flake

reiinakano commented on sklearn/model_selection/_validation.py:
```diff
@@ -727,12 +727,35 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+        if n_classes != len(estimator.classes_):
+            if predictions.ndim == 2 and \
+                    predictions.shape[1] != len(estimator.classes_):
+                # This handles the case when the shape of predictions
```
Actually, for the first case (ovo), the recommendation in the error message is helpful. We *can* handle the ovo case but only if the folds are stratified.

reiinakano committed:
wrong classes range

lesteve commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1259,7 +1339,8 @@ def test_cross_val_predict_class_subset():
         # Testing unordered labels
-        y = [1, 1, -4, 6]
+        y = np.array([x//20 for x in range(-100, 100, 2)])
+        y = shuffle(y, random_state=0)
```
or just `y = np.repeat([1, 1, -4, 6], 25)`. Always a good idea to avoid unnecessary compilcations ...

lesteve reviewed (commented):
> I think we still need a warning in any case when n_classes_training != n_classes.

You have not tackled this comment AFAICT. It would be good to add a test for this with assert_warns as well.

reiinakano committed:
address comments. adjust error message

reiinakano committed:
add warning

reiinakano committed:
change warning to runtimewarning

reiinakano committed:
add test for the warning

lesteve commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -776,6 +777,87 @@ def split(self, X, y=None, groups=None):
     assert_raises(ValueError, cross_val_predict, est, X, y, cv=BadCV())
 
+    X, y = load_iris(return_X_y=True)
+    assert_warns(RuntimeWarning, cross_val_predict, LogisticRegression(),
```
Can you use assert_warns_message to test the warning message?

jnothman replied:
Apart from this do you have any concerns with the PR?

lesteve replied:
I don't think so, I am fine with the general strategy as I mentioned above. There are some minor things I will fix shortly.

If you haven't already, maybe you can have a quick look at the warnings/error message and make sure they are fine.

lesteve replied:
Another thing: are you fine with float_min as the default value for missing classes for predict_log_proba and decision_function. Another option is `-inf` but maybe some consumers downstream will raise an error will infinite values.

jnothman replied:
predict_log_proba should probably use -inf... decision_function is a messy case.

jnothman replied:
I don't think float-min is perfect, but I'm documenting it at least.

lesteve committed:
Use assert_warns_message rather than assert_warns

Other minor fixes

lesteve commented on sklearn/model_selection/_validation.py:
```diff
@@ -727,12 +727,48 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+                    # it with. This case is found when sklearn.svm.SVC is
+                    # set to `decision_function_shape='ovo'`.
+                    raise ValueError('Output shape {} of {} does not match '
+                                     'number of classes ({}) in fold. Cannot '
```
I think this error message is wrong:
* I would remove "Cannot reconcile different number of classes in different folds, because the problem is specific to one fold
* the recommendation of using of using properly stratified folds does not apply. If we are sure that this problem only exists for SVC with `decision_shape_function='ovo'`, we should say use `decision_shape_function='ovr'` maybe?

amueller replied:
Maybe. It could happen in other situations? Could also be a custom estimator?

jnothman replied:
Or just say that wacky decision functions are not supported :)

jnothman commented:
I would like to see this merged and 0.19.1 released by the end of the week. Is that reasonable, @amueller?

jnothman committed:
Note on class-absent replacement values

jnothman committed:
Improve error message

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -625,6 +625,15 @@ def cross_val_predict(estimator, X, y=None, groups=None, cv=None, n_jobs=1,
+    default score needs to be assigned to all instances for that class if
+    ``method`` produces columns per class, as in {'decision_function',
+    'predict_proba', 'predict_log_proba'}.  For ``predict_proba`` this value is
+    0.  In order to ensure finite output, we approximate negative infinity by
```
I think this behaviour and note is sufficient for a first release.
